feat: support mulitple temporal filters in AdhocFilter and move the Time Section away

zhaoyongjie opened the pull request:
#### Background
The Time Section is in the Superset Explore now. The original design is for selecting Druid time partition and time grains.
Over time, Superset has supported more databases. The time filter should be a kind of Adhoc Filters.

#### Summary
This PR will move functionality of the Time Section to the Adhoc Filters when `GENERIC_CHART_AXES` is enabled, and then Adhoc Filters should support multiple time columns.

The new operator in the Adhoc filter that is `TEMPORAL_BETWEEN` has been added, so the new `queryObject` will support like this

```
{
  ...
  filters: {
    "col": "Order Date",
    "op": "TEMPORAL_RANGE",
    "val": "DATEADD(DATETIME(\"2022-10-10T00:00:00\"), -20, year) : 2022-10-10T00:00:00"
  }
  ...
}
```

the filter will generate a SQL snippet

```SQL
...
WHERE "Order Date" >= '2002-10-10 00:00:00.000000'
  AND "Order Date" < '2022-10-10 00:00:00.000000'
...
```

### Unsupported Charts
- Legacy Rose
- Legacy Pivot table
- Legacy Area
- Legacy Bar
- Legacy Compare
- Legacy DualLine
- Legacy Line
- Legacy TimePivot
- TimeTable


![image](https://user-images.githubusercontent.com/2016594/196891391-51ba2535-50cf-48a4-bced-50b7ae4afb36.png)



https://user-images.githubusercontent.com/2016594/196890988-57327f9f-a2df-4cda-8269-2550a194db4b.mov




### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

zhaoyongjie committed:
feat: support mulitple datetime filter in AdhocFilter and move away time section

zhaoyongjie committed:
should pass all CI steps

zhaoyongjie committed:
add UT for time_range_utils

zhaoyongjie committed:
fix time offset

villebro commented on superset-frontend/packages/superset-ui-core/test/query/buildQueryContext.test.ts:
```diff
@@ -18,6 +18,7 @@
  */
 import { buildQueryContext } from '@superset-ui/core';
 import * as queryModule from '../../src/query/normalizeTimeColumn';
+import * as getXAxisModule from '../../src/query/getXAxis';
```
I know this is following the same convention as on the line above, but I personally prefer to avoid `import *` whenever possible.

zhaoyongjie replied:
The module is used to mock an object, so we have to do `import * as moduleName from 'reletive path'` statement here.

villebro replied:
Oh yes, makes sense 👍

villebro reviewed (commented):
First pass comments. Super stoked to finally see the time section being removed! Also, it would be really great if we could have some design eyes on the adhoc filter modal, as the time picker pill feels slightly sketchy in this particular context (even making the pill full width would be an improvement IMO)

zhaoyongjie commented on superset-frontend/src/explore/constants.ts:
```diff
@@ -45,6 +45,7 @@ export enum Operators {
   LATEST_PARTITION = 'LATEST_PARTITION',
   IS_TRUE = 'IS_TRUE',
   IS_FALSE = 'IS_FALSE',
+  DATETIME_BETWEEN = 'DATETIME_BETWEEN',
```
Good catch, I will change it to `TEMPORAL_BETWEEN`.

villebro replied:
If I have an open ended range, like "100 years ago", then isn't it rather a "BEFORE" than a "BETWEEN"? So I would still reconsider renaming this to just "TEMPORAL" or something else instead of "BETWEEN".

zhaoyongjie replied:
the `TEMPORAL` sounds like a date type rather than an operator. how about `TEMPORAL_FROM` or `TEMPORAL_PARSER` or `CACULATE_TEMPORAL`

villebro replied:
How do you feel about `TEMPORAL_RANGE`?

zhaoyongjie replied:
Like this. Thanks to InteliJ, renaming is done in 1 second.

zhaoyongjie committed:
fix pylint

zhaoyongjie commented on superset-frontend/packages/superset-ui-core/src/query/buildQueryContext.ts:
```diff
@@ -48,15 +53,26 @@ export default function buildQueryContext(
       // eslint-disable-next-line no-param-reassign
       query.post_processing = query.post_processing.filter(Boolean);
     }
+    if (hasGenericChartAxes && query.time_range) {
+      // eslint-disable-next-line no-param-reassign
+      query.filters = ensureIsArray(query.filters).map(flt =>
+        flt?.op === 'DATETIME_BETWEEN'
+          ? ({ ...flt, val: query.time_range } as BinaryQueryObjectFilterClause)
+          : flt,
+      );
+    }
```
Yes, it's for applying native filters to the ad-hoc filter on the Dashboard, this part will override all ad-hoc time filters on the Dashboard.

In the original design is that the chart only has a single time filter so the native filter was designed only to pass `time_range` to every chart rather than pass `time column`(granularity in the QueryObject) and `time value`(time_range in the QueryObject) to the chart.

To solve this issue, I think we might need to specify the time columns in the native filter so that the user knows which time column would apply.

zhaoyongjie commented on superset/db_engine_specs/sqlite.py:
```diff
@@ -77,7 +77,11 @@ def convert_dttm(
     ) -> Optional[str]:
         tt = target_type.upper()
-        if tt in (utils.TemporalType.TEXT, utils.TemporalType.DATETIME):
+        if tt in (
+            utils.TemporalType.TEXT,
+            utils.TemporalType.DATETIME,
+            utils.TemporalType.TIMESTAMP,
+        ):
```
by catch: SQLite also has a timestamp type.

zhaoyongjie committed:
datetime_between => temporal_between

zhaoyongjie committed:
add more UT

codyml reviewed (commented):
Not doing a full review because most I'm pretty out of my depth on most of the code changes, but I tested locally on several ECharts and it seems to work great!  I did see a few issues, but they were also present on master so not sure if they're relevant here:
- "Original Value" option in time grain is a little buggy for me, if you select it the selection disappears but if you select it again it appears.
- For charts with "Temporal X-axis", you can't remove the default item with the little "x".  If there are two options and you choose one that's not default, then click the "x", it'll just go back to the original one.
- Some charts had a tooltip value that seemed off: 
<img width="700" alt="Screen Shot 2022-10-18 at 7 09 52 PM" src="https://user-images.githubusercontent.com/13007381/196579233-05dc61ba-c333-4520-9c07-769829a9109e.png">

My one other comment is that checking feature flags at module execution time has been responsible for some weird bugs in the past due to race conditions between when they're set and read (that's what #21315 was fixing). It would be great if we could only check feature flags during React rendering or other execution points that are guaranteed to happen after initial module execution. That honestly looks pretty hard to do in this PR, but just wanted to note that this might be a source of bugs in the future.

zhaoyongjie committed:
fix mysql failure UT

zhaoyongjie commented:
> "Original Value" option in time grain is a little buggy for me, if you select it the selection disappears but if you select it again it appears.
Yes, It's a bug because the `Original value` and unselected shared the same key(`null`) in the selection control. The easiest way to fix this is disabled removing the value in the time grain selection. I will fix that in the separate PR because it's also on the master branch.

> For charts with "Temporal X-axis", you can't remove the default item with the little "x". If there are two options and you choose one that's not default, then click the "x", it'll just go back to the original one.
What I'm trying to say is, this is all original Druid logic. There is no default time column in the general-purpose database or general BI. It is a little weird design to always place a time column for a generic axis control (which is not a temporal x-axis). this is only my personal view, welcome to discuss.

> Some charts had a tooltip value that seemed off
Good catch, I will check it out.

> checking feature flags
Feature flags are global static constant for the frontend codebase, these values can't be reset by the frontend codes because they're Flask application context config items, in the other words, the only way to change these configs that need to restart the Superset process.

zhaoyongjie committed:
new style in date picker

zhaoyongjie committed:
datepicker modal in adhoc filter

zhaoyongjie committed:
fix overlap in modal

zhaoyongjie committed:
fix evalResponse first open modal

zhaoyongjie committed:
add header

zhaoyongjie commented:
/testenv up FEATURE_GENERIC_CHART_AXES=true FEATURE_DRILL_TO_DETAIL=true FEATURE_DASHBOARD_CROSS_FILTERS=true

villebro commented on superset/common/query_context_processor.py:
```diff
@@ -314,30 +316,49 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
         time_offsets = query_object.time_offsets
-        outer_from_dttm = query_object.from_dttm
-        outer_to_dttm = query_object.to_dttm
+        outer_from_dttm, outer_to_dttm = get_since_until_from_query_object(query_object)
+        if not outer_from_dttm or not outer_to_dttm:
+            raise QueryObjectValidationError(
+                _(
+                    "An enclosed time range (both start and end) must be specified "
+                    "when using a Time Comparison."
+                )
+            )
```
Out of scope for this PR, but I believe we could remove this requirement, as I feel we should be able to do time comparison without specifying a time range (I usually work around this by placing a `100 years ago : now` on the time filter)

villebro commented on superset/common/query_object.py:
```diff
@@ -336,6 +335,7 @@ def to_dict(self) -> Dict[str, Any]:
             "series_limit": self.series_limit,
             "series_limit_metric": self.series_limit_metric,
             "to_dttm": self.to_dttm,
+            "time_shift": self.time_shift,
```
I believe this will invalidate all currently stored cache values: to avoid cache misses, I've previously placed a check on `QueryObject.cache_key` checking if the new property is truthy, and if so, add a new key to the cache key dict (the idea is to move all of these in one go to `QueryObject.to_dict()` to make a big bang during e.g. a major release).

zhaoyongjie replied:
`time_shift` is not really useful in existing code. It seems to be a parameter used by the old time picker, and I'm only using it here to ensure backward compatibility for the query object.

villebro commented on superset/models/helpers.py:
```diff
@@ -1348,6 +1348,7 @@ def get_sqla_query(  # pylint: disable=too-many-arguments,too-many-locals,too-ma
         row_offset: Optional[int] = None,
         timeseries_limit: Optional[int] = None,
         timeseries_limit_metric: Optional[Metric] = None,
+        time_shift: Optional[str] = None,
```
I'm not familiar with these helpers, but do we need to also copy over the new logic from `get_sqla_query()` in `connectors/sqla/models.py`? Or better yet, remove one of these to avoid code duplication?

zhaoyongjie replied:
IMO, This helper has to do full refactoring. otherwise, all our code would have to be duplicated. the `time_shift` here is just to prevent raising an error from function calls.

villebro reviewed (commented):
Looking good - did pretty extensive testing with and without the feature flag, both in Explore and Dashboard view with native time filters, and everything seems to work great! A few last comments, but feeling ready to approve after this round.

zhaoyongjie committed:
apply global default time filter

zhaoyongjie committed:
temporal_between => temporal_range

zhaoyongjie committed:
code smell

zhaoyongjie committed:
fix drag&drop and save temporal column

zhaoyongjie committed:
fix subject change

zhaoyongjie committed:
fix adhoc filter title

zhaoyongjie committed:
use isTemporalColumn

zhaoyongjie committed:
add ut

zhaoyongjie commented:
/testenv up FEATURE_GENERIC_CHART_AXES=true FEATURE_DRILL_TO_DETAIL=true FEATURE_DASHBOARD_CROSS_FILTERS=true

zhaoyongjie committed:
fix previous UT

zhaoyongjie committed:
fix ut

zhaoyongjie committed:
support Handlebars

zhaoyongjie committed:
ut for time filter

villebro reviewed (commented):
Really nice work! 👏  I don't really have anything to comment on the code, and the test coverage is awesome! 💪  However, one thing that I noticed is that `DEFAULT_TIME_FILTER` doesn't apply by default when entering explore - you have to add a temporal filter, and only then will it apply. This will likely cause perf issues on deployments with massive data warehouses where the default time filter is used to avoid big table scans.

A potential fix for this that we might want to consider: when Exploring a dataset (=creating a new chart), should we always add a new time filter if there is a temporal column in the dataset, similar to how it now works in the legacy time section? We may even want to do that even if the default filter is set to "No filter" to be able to apply native time filters to these charts. It may look weird to have the time filter there by default, but IMO it may be needed to ensure the functionality stays unchanged from the current one.

zhaoyongjie committed:
addressing comments

zhaoyongjie committed:
fix lint

zhaoyongjie commented:
HI @villebro, I have addressed the comments that should populate the default time column into Adhoc Filters and the `DEFAULT_TIME_FILTER` could be set from `config.py`. I'd really appreciate your eyes on the last review codes when you have time.

IMO, the default column in the Dataset and the default filter value might be redesigned so that it supports the non-temporal column as the default filter.


https://user-images.githubusercontent.com/2016594/198513874-e1bc505d-e5c9-4f8f-a68e-23773b1bc8b0.mov

villebro reviewed (approved):
LGTM. Comparing the old UX with the new one, I think this new behavior is much more intuitive than the last one + this brings the much anticipated functionality of multiple time filters with a very intuitive UX flow. Thanks for the incredible effort on this one, this is a major milestone in finally shedding some of the last and most annoying remnants of the legacy Druid connector! ❤️ 🚀

zhaoyongjie commented:
> LGTM. Comparing the old UX with the new one, I think this new behavior is much more intuitive than the last one + this brings the much anticipated functionality of multiple time filters with a very intuitive UX flow. Thanks for the incredible effort on this one, this is a major milestone in finally shedding some of the last and most annoying remnants of the legacy Druid connector! ❤️ 🚀

Thanks so much for the review. You make my day.

villebro commented:
> IMO, the default column in the Dataset and the default filter value might be redesigned so that it supports the non-temporal column as the default filter.

Supporting non-temporal default filters is an excellent idea! It may of course be slightly more difficult to define a globally applicable non-temporal default filter (what would the column be?), but this should at least be possible to define on the dataset level.

zhaoyongjie committed:
fix ut

jinghua-qa commented:
/testenv up FEATURE_GENERIC_CHART_AXES=true FEATURE_DRILL_TO_DETAIL=true FEATURE_DASHBOARD_CROSS_FILTERS=true

zhaoyongjie committed:
fix label error when custom SQL => simple panel

zhaoyongjie commented:
/testenv up FEATURE_GENERIC_CHART_AXES=true FEATURE_DRILL_TO_DETAIL=true FEATURE_DASHBOARD_CROSS_FILTERS=true

zhaoyongjie committed:
ut for useDatePickerInAdhocfilter

jinghua-qa commented:
Thank you for the great work~
I have a question, it looks like when i use custom sql in filter with time column, time comparison did not support, is this expected?
Steps:
1, add a time column to filter
2, use time comparison to define start and end
3, go to time comparison and select compare time

Expect:
time comparison will work with time filter

Actual:
time comparison did not work with time filter

https://user-images.githubusercontent.com/81597121/199279123-f07f150d-7db2-4149-ae2d-1e87f9253451.mov

zhaoyongjie commented:
> Thank you for the great work~ I have a question, it looks like when i use custom sql in filter with time column, time comparison did not support, is this expected? Steps: 1, add a time column to filter 2, use time comparison to define start and end 3, go to time comparison and select compare time

Hi @jinghua-qa, the `Time Comparision` + `Custom SQL` doesn't been supported yet. Currently, the `Time Comparision` should with `time range` in the Adhoc filter rather than `Custom SQL`.

jinghua-qa commented:
> > Thank you for the great work~ I have a question, it looks like when i use custom sql in filter with time column, time comparison did not support, is this expected? Steps: 1, add a time column to filter 2, use time comparison to define start and end 3, go to time comparison and select compare time
> 
> Hi @jinghua-qa, the `Time Comparision` + `Custom SQL` doesn't been supported yet. Currently, the `Time Comparision` should with `time range` in the Adhoc filter rather than `Custom SQL`.

Are we going to support this feature in the future? it looks like when we move the time range to filter, user can use custom sql to support time column as filter, it will be great if we can support that with time comparison.

zhaoyongjie commented:
> > > Thank you for the great work~ I have a question, it looks like when i use custom sql in filter with time column, time comparison did not support, is this expected? Steps: 1, add a time column to filter 2, use time comparison to define start and end 3, go to time comparison and select compare time
> > 
> > 
> > Hi @jinghua-qa, the `Time Comparision` + `Custom SQL` doesn't been supported yet. Currently, the `Time Comparision` should with `time range` in the Adhoc filter rather than `Custom SQL`.
> 
> Are we going to support this feature in the future? it looks like when we move the time range to filter, user can use custom sql to support time column as filter, it will be great if we can support that with time comparison.


If we want to support `custom SQL` + `Time Comparison`, there are some pre-requirement works, e.g. 1) parsing and validating it. 2) it should contain the start time and end time. This part of the work is completely beyond the scope of this PR.

jinghua-qa reviewed (approved):
LGTM

korpa commented:
Two questions on this feature:

1) How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns

2) How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?

bilalabdullah44000 commented:
> Two questions on this feature:
> 
> 1. How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns
> 2. How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?



Any update on this ??

zhaoyongjie commented:
> Two questions on this feature:
> 
> 1. How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns
What version of Superset do you use? Native filter in the Dashboard should append a new time filter, so I think it should be worked on Dashboard. 


> 2. How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?

Should support Jinja Template if not, it should be a regression bug.

bilalabdullah44000 commented:
> > Two questions on this feature:
> > 
> > 1. How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns
> 
> What version of Superset do you use? Native filter in the Dashboard should append a new time filter, so I think it should be worked on Dashboard.
> 
> > 2. How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?
> 
> Should support Jinja Template if not, it should be a regression bug.


1) In jinja, it is giving result in the form of
                {
                  ...
                  filters: {
                    "col": "Date1",
                    "op": "TEMPORAL_RANGE",
                    "val": "DATEADD(DATETIME(\"2022-10-10T00:00:00\"), -20, year) : 2022-10-10T00:00:00"
                  },
                  {
                    "col": "Date1",
                    "op": "TEMPORAL_RANGE",
                    "val": "Last day : 2022-10-10T00:00:00"
                  }
                  
                }
                
                ideally it should give result in the form of timestamp so it can be easily used in sql just like from_dttm and to_dttm. 
                Can we parse above values to get timestamp from above format effectively ?

2) Also at my end, if i create two dashboard filters i.e. created_at and updated_at , it is applying filter values against these 2 filters on my default datetime/temporal column i.e. last_state_change_time 

![Screenshot from 2023-07-07 17-09-01](https://github.com/apache/superset/assets/111341070/5c9cbb55-c986-42c6-ac8e-caf6d8515368)

zhaoyongjie commented:
> > > Two questions on this feature:
> > > 
> > > 1. How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns
> > 
> > 
> > What version of Superset do you use? Native filter in the Dashboard should append a new time filter, so I think it should be worked on Dashboard.
> > > 2. How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?
> > 
> > 
> > Should support Jinja Template if not, it should be a regression bug.
> 
> 1. In jinja, it is giving result in the form of
>    {
>    ...
>    filters: {
>    "col": "Date1",
>    "op": "TEMPORAL_RANGE",
>    "val": "DATEADD(DATETIME("2022-10-10T00:00:00"), -20, year) : 2022-10-10T00:00:00"
>    },
>    {
>    "col": "Date1",
>    "op": "TEMPORAL_RANGE",
>    "val": "Last day : 2022-10-10T00:00:00"
>    }
>    ```
>             }
>    ```
> 
> ideally it should give result in the form of timestamp so it can be easily used in sql just like from_dttm and to_dttm. Can we parse above values to get timestamp from above format effectively ?

The **val** in each time filter means that dynamically calculate "from date time" and "to date time". If you just want to use fixed timestamp, leave timestamp string in Custom, for instance:

<img width="991" alt="image" src="https://github.com/apache/superset/assets/2016594/367a638c-a08a-4acd-b617-8f2fe03c21b6">



> 
> 2. Also at my end, if i create two dashboard filters i.e. created_at and updated_at , it is applying filter values against these 2 filters on my default datetime/temporal column i.e. last_state_change_time
> 
> ![Screenshot from 2023-07-07 17-09-01](https://user-images.githubusercontent.com/111341070/251736265-5c9cbb55-c986-42c6-ac8e-caf6d8515368.png)


It might be a regression bug, I've revert this [PR](https://github.com/apache/superset/pull/24176) in my custom Superset. I hope it should help with you.

zhaoyongjie commented:
@bilalabdullah44000 posted an incorrect link, just updated

bilalabdullah44000 commented:
> > > > Two questions on this feature:
> > > > 
> > > > 1. How does this work in dashboards. Even if I add two time range filters in a dashboard, it is not possible to select the correct date columns
> > > 
> > > 
> > > What version of Superset do you use? Native filter in the Dashboard should append a new time filter, so I think it should be worked on Dashboard.
> > > > 2. How does this feature work with jinja templates? Are there any other variables than `{{ from_dttm }}` and `{{ to_dttm }}` to support this feature?
> > > 
> > > 
> > > Should support Jinja Template if not, it should be a regression bug.
> > 
> > 
> > 
> > 1. In jinja, it is giving result in the form of
> >    {
> >    ...
> >    filters: {
> >    "col": "Date1",
> >    "op": "TEMPORAL_RANGE",
> >    "val": "DATEADD(DATETIME("2022-10-10T00:00:00"), -20, year) : 2022-10-10T00:00:00"
> >    },
> >    {
> >    "col": "Date1",
> >    "op": "TEMPORAL_RANGE",
> >    "val": "Last day : 2022-10-10T00:00:00"
> >    }
> >    ```
> >             }
> >    ```
> > 
> > ideally it should give result in the form of timestamp so it can be easily used in sql just like from_dttm and to_dttm. Can we parse above values to get timestamp from above format effectively ?
> 
> The **val** in each time filter means that dynamically calculate "from date time" and "to date time". If you just want to use fixed timestamp, leave timestamp string in Custom, for instance:
> 
> <img alt="image" width="991" src="https://user-images.githubusercontent.com/2016594/251737774-367a638c-a08a-4acd-b617-8f2fe03c21b6.png">
> > 2. Also at my end, if i create two dashboard filters i.e. created_at and updated_at , it is applying filter values against these 2 filters on my default datetime/temporal column i.e. last_state_change_time
> > 
> > ![Screenshot from 2023-07-07 17-09-01](https://user-images.githubusercontent.com/111341070/251736265-5c9cbb55-c986-42c6-ac8e-caf6d8515368.png)
> 
> It might be a regression bug, I've revert this [PR](https://github.com/apache/superset/pull/23021) in my custom Superset. I hope it should help with you.



I Understand. if i select timecolumn , then the time range filter will be applied on the column selected in the time column. but how do i apply multiple timerange filters i.e. filter on both created_by column and updated_by column at dashboard level ?

bilalabdullah44000 commented:
@zhaoyongjie Can we apply multiple time range filters at the same time on dashboard as well like slice explore ?

zhaoyongjie commented:
> @zhaoyongjie Can we apply multiple time range filters at the same time on dashboard as well like slice explore ?

IMO, Dashboard must be applied multiple time range filter if not it's a bug.

bilalabdullah44000 commented:
> > @zhaoyongjie Can we apply multiple time range filters at the same time on dashboard as well like slice explore ?
> 
> IMO, Dashboard must be applied multiple time range filter. If not, it's a bug.

Here in attached screenshot, I have applied two time range filters i.e. created_at and updated_at.

As seen in the query, Only one filter is applied on default temporal column or on the column which is selected in time column field in dashboard filters section. 

![251736265-5c9cbb55-c986-42c6-ac8e-caf6d8515368](https://github.com/apache/superset/assets/111341070/de090260-2206-44e5-b695-31ec82b0680a)

zhaoyongjie commented:
@bilalabdullah44000 You need to use Time Column filter and Time Range filter together to select multiple time filter, but I also found this feature was not correct work on the master, the Time Range replaced all of the value of time filters. I suggest that the first step revert the PR https://github.com/apache/superset/pull/23021 and then try to fix it in your side.


<img width="1398" alt="image" src="https://github.com/apache/superset/assets/2016594/d311b2ca-fcc6-46e4-9802-d79189435d70">

zhaoyongjie commented:
> @bilalabdullah44000 You need to use Time Column filter and Time Range filter together to select multiple time filter, but I also found this feature was not correct work on the master, the Time Range replaced all of the value of time filters. I suggest that the first step revert the PR #23021 and then try to fix it in your side.
> 
> <img alt="image" width="1398" src="https://user-images.githubusercontent.com/2016594/252954541-d311b2ca-fcc6-46e4-9802-d79189435d70.png">

By this feature, I think we should redesign the Time Filter on the Dashboard, should merge the Time Column/Time Range/Time Grain together as a unique Time Filter. 

Unfortunately, the new Time Filter almost didn't support multiple filters on the Dashboard, and some codebase was broken previous design, ---- the Native Filter should **APPEND** the new _where clause_ rather than **REPLACE** value of filter of chart.
